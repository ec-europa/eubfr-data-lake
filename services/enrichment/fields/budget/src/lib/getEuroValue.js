// @flow
import request from 'request-promise-native';
import type { BudgetItem, TimePrecision } from '@eubfr/types';

const padDate = (date: number): number | string =>
  date < 10 ? `0${date}` : date;

export const getEuroValue = async (
  inputBudgetItem: BudgetItem,
  projectEndDate: Date,
  projectEndPrecision: TimePrecision
) => {
  let euroValue = null;
  const precision = 100; // round converted value to the nearest 100

  // Least precision: year
  let period = projectEndDate.getFullYear();
  let apiPeriod = 'A';

  // Best precision: day
  if (projectEndPrecision === 'day') {
    const month = projectEndDate.getMonth() + 1;
    const day = projectEndDate.getDate();
    period = `${period}-${padDate(month)}-${padDate(day)}`;
    apiPeriod = 'D';
  }

  // If not a day, try at least: month
  else if (projectEndPrecision === 'month') {
    const month = projectEndDate.getMonth() + 1;
    period = `${period}-${padDate(month)}`;
    apiPeriod = 'M';
  }

  // "EXR/A.[currency].EUR.SP00.A" corresponds to the key of a dataset
  // The keys can be found here: http://sdw.ecb.europa.eu/browseTable.do?node=1495
  // API info: https://sdw-wsrest.ecb.europa.eu/web/generator/index.html
  const url = `https://sdw-wsrest.ecb.europa.eu/service/data/EXR/${apiPeriod}.${
    inputBudgetItem.currency
  }.EUR.SP00.A`;

  let results;
  const qs = {
    startPeriod: period,
    endPeriod: period,
    detail: 'dataonly',
  };

  try {
    results = await request.get({
      url,
      qs,
      json: true,
      headers: {
        Accept: 'application/vnd.sdmx.data+json;version=1.0.0-wd',
      },
    });
  } catch (e) {
    console.error('error', url, qs, e);
    return euroValue;
  }

  if (
    results &&
    results.dataSets &&
    results.dataSets[0] &&
    results.dataSets[0].series &&
    results.dataSets[0].series['0:0:0:0:0'] &&
    results.dataSets[0].series['0:0:0:0:0'].observations &&
    results.dataSets[0].series['0:0:0:0:0'].observations[0]
  ) {
    const exr = results.dataSets[0].series['0:0:0:0:0'].observations['0'][0];
    euroValue = Math.ceil(inputBudgetItem.value / exr / precision) * precision;
  }

  // Fallback request decreasing precision from daily to monthly
  if (results === undefined && projectEndPrecision === 'day') {
    euroValue = await getEuroValue(inputBudgetItem, projectEndDate, 'month');
  }

  return euroValue;
};

export default getEuroValue;
