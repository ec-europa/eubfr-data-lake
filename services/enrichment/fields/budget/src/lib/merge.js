/* eslint-disable no-underscore-dangle */
export const mergeRecords = (newRecord, existingRecord) => {
  if (!newRecord) return null;

  const mergedRecord = JSON.parse(JSON.stringify(newRecord));

  // What we need to retrieve from the existing record
  mergedRecord.computed_key = existingRecord.computed_key;

  // Budget: copy existing record if already enriched
  const budgetFields = [
    'eu_contrib',
    'other_contrib',
    'private_fund',
    'public_fund',
    'total_cost',
  ];

  if (existingRecord.budget && newRecord.budget) {
    budgetFields.forEach(field => {
      if (
        existingRecord.budget[field] &&
        mergedRecord.budget[field] &&
        existingRecord.budget[field]._original &&
        existingRecord.budget[field]._original.currency ===
          mergedRecord.budget[field].currency &&
        existingRecord.budget[field]._original.value ===
          mergedRecord.budget[field].value
      ) {
        mergedRecord.budget[field] = JSON.parse(
          JSON.stringify(existingRecord.budget[field])
        );
      }
    });
  }

  return mergedRecord;
};

export default mergeRecords;
