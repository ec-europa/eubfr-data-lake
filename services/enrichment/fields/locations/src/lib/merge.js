export const mergeRecords = (newRecord, existingRecord) => {
  if (!newRecord) return null;

  const mergedRecord = JSON.parse(JSON.stringify(newRecord));

  if (existingRecord) {
    // extract enriched data if available
  }

  return mergedRecord;
};

export default mergeRecords;
