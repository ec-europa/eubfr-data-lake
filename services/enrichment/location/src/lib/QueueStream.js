import stream from 'stream';

export default class QueueStream extends stream.Writable {
  constructor(options) {
    super(options);
    this.sqs = options.sqs;
    this.queueUrl = options.queueUrl;
  }

  _write(chunk, enc, next) {
    return this.sqs.sendMessage(
      {
        MessageAttributes: {},
        MessageBody: JSON.stringify(chunk),
        QueueUrl: this.queueUrl,
      },
      err => {
        if (err) {
          console.log('Error', err);
        }

        next();
      }
    );
  }
}
