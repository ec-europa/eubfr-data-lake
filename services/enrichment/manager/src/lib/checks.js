export const canLocationsBeEnriched = locations =>
  locations &&
  locations.some(loc => {
    if (!loc) {
      return false;
    }

    // Centroid is empty but country code is provided
    const needsLocation = !loc.centroid && loc.country_code;
    if (needsLocation) return true;

    // Country code is empty but centroid is provided
    const needsCountryCode = !loc.country_code && loc.centroid;

    if (needsCountryCode) return true;

    return false;
  });

export const needsEnrichment = record =>
  record &&
  record.project_locations &&
  canLocationsBeEnriched(record.project_locations);
