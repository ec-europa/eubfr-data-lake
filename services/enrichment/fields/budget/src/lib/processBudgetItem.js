// @flow
import sanitizeBudgetItem from '@eubfr/lib/budgetFormatter';
import type { BudgetItem } from '@eubfr/types';

import { getEuroValue } from './getEuroValue';

// List of currencies that can be translated into EUR
// Manually extracted form ECB's API
// http://sdw.ecb.europa.eu/browseTable.do?node=1495 ("Currency" field)
const availableCurrencies = [
  'AUD',
  'BGN',
  'BRL',
  'CAD',
  'CHF',
  'CNY',
  'CYP',
  'CZK',
  'DKK',
  'DZD',
  'EEK',
  'EGP',
  'GBP',
  'GRD',
  'HKD',
  'HRK',
  'HUF',
  'IDR',
  'ILS',
  'INR',
  'ISK',
  'JPY',
  'KRW',
  'LTL',
  'LVL',
  'MAD',
  'MTL',
  'MXN',
  'MYR',
  'NOK',
  'NZD',
  'PHP',
  'PLN',
  'RON',
  'RUB',
  'SEK',
  'SGD',
  'SIT',
  'SKK',
  'THB',
  'TRY',
  'TWD',
  'USD',
  'ZAR',
];

export const processBudgetItem = async (
  inputBudgetItem: BudgetItem,
  projectEnd: string,
  projectEndPrecision: string
) => {
  if (!inputBudgetItem) return null;

  // check if it at least one is in EUR
  if (inputBudgetItem.currency === 'EUR') return inputBudgetItem;

  // ensure date is provided, as it's required
  if (!projectEnd) return inputBudgetItem;

  if (
    inputBudgetItem.currency &&
    inputBudgetItem.currency !== 'EUR' &&
    availableCurrencies.indexOf(inputBudgetItem.currency) >= 0
  ) {
    const projectEndDate = new Date(projectEnd);

    const euroValue = await getEuroValue(
      inputBudgetItem,
      projectEndDate,
      projectEndPrecision
    );

    if (euroValue) {
      const formattedEuroBudget = sanitizeBudgetItem({
        value: euroValue,
        currency: 'EUR',
        raw: `EUR ${euroValue}`,
        _original: {
          value: inputBudgetItem.value,
          currency: inputBudgetItem.currency,
          raw: inputBudgetItem.raw,
        },
      });

      // Return enriched record
      return formattedEuroBudget;
    }
  }

  return inputBudgetItem;
};

export default processBudgetItem;
