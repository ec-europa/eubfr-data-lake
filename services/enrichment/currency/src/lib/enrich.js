import { mergeRecords } from './merge';
// import { enrichCurrency } from '../plugins/currency';

export const enrich = async (record, existingRecord) => {
  /* if (!record.project_locations || record.project_locations.length === 0) {
    return record;
  } */

  const enrichedRecord = mergeRecords(record, existingRecord);

  // DO THE WORK
  console.log('ENRICH CURRENCY');

  return enrichedRecord;
};

export default enrich;
