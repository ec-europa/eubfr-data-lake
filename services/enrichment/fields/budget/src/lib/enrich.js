/**
 * Convert foreign currencies into EUR usign the exchange rate at project end
 * https://webgate.ec.europa.eu/CITnet/jira/browse/EUBFR-184
 */

import { processBudgetItem } from './processBudgetItem';

export const enrich = async record => {
  if (!record) return null;

  const enrichedRecord = JSON.parse(JSON.stringify(record));

  // If project end date or the budget is not provided
  // or if '_original' is already present,
  // then skip enrichment
  if (
    !enrichedRecord.timeframe ||
    !enrichedRecord.timeframe.to ||
    !enrichedRecord.budget ||
    enrichedRecord.budget._original
  ) {
    return enrichedRecord;
  }

  const budgetFields = [
    'eu_contrib',
    'other_contrib',
    'private_fund',
    'public_fund',
    'total_cost',
  ];

  const enrichedBudgetFields = await Promise.all(
    budgetFields.map(async field =>
      processBudgetItem(
        enrichedRecord.budget[field],
        enrichedRecord.timeframe.to,
        enrichedRecord.timeframe.to_precision
      )
    )
  );

  // Enrich here
  enrichedBudgetFields.forEach((field, index) => {
    enrichedRecord.budget[budgetFields[index]] = field;
  });

  return enrichedRecord;
};

export default enrich;
