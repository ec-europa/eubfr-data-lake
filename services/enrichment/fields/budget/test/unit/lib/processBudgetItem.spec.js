/**
 * @jest-environment node
 */

import { processBudgetItem } from '../../../src/lib/processBudgetItem';

describe('processBudgetItem helper function', () => {
  test('Returns null when inputBudgetItem is not provided', async () => {
    const result = await processBudgetItem();
    expect(result).toBeNull();
  });

  test('Returns original budget object if currency is EUR (nothing to enrich)', async () => {
    const budgetItem = {
      value: 63115,
      currency: 'EUR',
      raw: '63,115.00',
    };
    const result = await processBudgetItem(
      budgetItem,
      '2015-02-02T00:00:00.000Z'
    );
    expect(result).toMatchObject(budgetItem);
  });

  test('Returns original budget object if date is not provided', async () => {
    const budgetItem = {
      value: 63115,
      currency: 'BGN',
      raw: '63,115.00',
    };
    const result = await processBudgetItem(budgetItem);
    expect(result).toMatchObject(budgetItem);
  });

  test('Enriches non-EUR currency budget items when date is provided, but no precision, defaults to year precision', async () => {
    const budgetItem = {
      value: 63115,
      currency: 'BGN',
      raw: '63,115.00',
    };
    const result = await processBudgetItem(
      budgetItem,
      '2015-02-02T00:00:00.000Z'
    );
    // Assert convertion to EUR
    expect(result.currency).toBe('EUR');
    // Assert change in original value
    expect(result.value).not.toBe(budgetItem.value);
    // Assert structure in overall
    expect(result).toMatchSnapshot();
  });

  test('Enrichment scenario, daily precision, exchange rate available', async () => {
    const budgetItem = {
      value: 63115,
      currency: 'BGN',
      raw: '63,115.00',
    };
    const result = await processBudgetItem(
      budgetItem,
      '2015-02-02T00:00:00.000Z',
      'day'
    );
    expect(result).toMatchSnapshot();
  });

  test('Enrichment scenario, daily precision, exchange rate not available, fallback to monthly', async () => {
    const budgetItem = {
      value: 63115,
      currency: 'USD',
      raw: '63,115.00',
    };
    const result = await processBudgetItem(
      budgetItem,
      '2014-12-25T00:00:00.000Z',
      'day'
    );
    expect(result).toMatchSnapshot();
  });

  test('Enrichment scenario, monthly precision', async () => {
    const budgetItem = {
      value: 63115,
      currency: 'BGN',
      raw: '63,115.00',
    };
    const result = await processBudgetItem(
      budgetItem,
      '2014-11-02T00:00:00.000Z',
      'month'
    );
    expect(result).toMatchSnapshot();
  });
});
