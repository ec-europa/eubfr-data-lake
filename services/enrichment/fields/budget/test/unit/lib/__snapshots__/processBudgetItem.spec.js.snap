// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processBudgetItem helper function Enriches non-EUR currency budget items when date is provided, but no precision, defaults to year precision 1`] = `
Object {
  "_original": Object {
    "currency": "BGN",
    "raw": "63,115.00",
    "value": 63115,
  },
  "currency": "EUR",
  "raw": "EUR 32300",
  "value": 32300,
}
`;

exports[`processBudgetItem helper function Enrichment scenario, daily precision, exchange rate available 1`] = `
Object {
  "_original": Object {
    "currency": "BGN",
    "raw": "63,115.00",
    "value": 63115,
  },
  "currency": "EUR",
  "raw": "EUR 32300",
  "value": 32300,
}
`;

exports[`processBudgetItem helper function Enrichment scenario, daily precision, exchange rate not available, fallback to monthly 1`] = `
Object {
  "_original": Object {
    "currency": "USD",
    "raw": "63,115.00",
    "value": 63115,
  },
  "currency": "EUR",
  "raw": "EUR 51200",
  "value": 51200,
}
`;

exports[`processBudgetItem helper function Enrichment scenario, monthly precision 1`] = `
Object {
  "_original": Object {
    "currency": "BGN",
    "raw": "63,115.00",
    "value": 63115,
  },
  "currency": "EUR",
  "raw": "EUR 32300",
  "value": 32300,
}
`;
